torch/csrc/distributed/c10d/nvshmem_extension.cu

```diff
@@ -311,11 +311,164 @@ at::Tensor nvshmem_all_to_all_vdev(
+  int bid = blockIdx.x;
+  int tid = threadIdx.x;
+
+  // Calculate the output offsets
+  __shared__ int64_t e_offsets[THREADS_PER_BLOCK];
+  prefixSum(e_offsets, output_splits, nsplits);
+  __syncthreads();
+
+  // Target a different e based on bid
+  for (int eid = bid; eid < nsplits; eid += gridDim.x) {
+    int peer = eid % npes;
+    // Amount from `peer` for `e`
+    auto peer_size = output_splits[eid] * stride;
+    auto source_offset = source_offsets[eid] * stride;
+    auto write_offset = e_offsets[eid] * stride;
+    nvshmemx_getmem_block(
+      (char*)recv_data + write_offset,
+      (char*)send_data + source_offset,
+      peer_size,
+      peer);
+  }
+  // Write out the output offsets (to the scratchpad line)
+  if (bid == 0 && tid < nsplits) {
+    source_offsets[tid] = e_offsets[tid];
+  }
+}
+
+at::Tensor nvshmem_all_to_all_vdev_2d(
+    at::Tensor& input,
+    at::Tensor& out,
+    at::Tensor& in_out_splits,
+    std::string group_name) {
+  /* Perform a 2D AllToAllv shuffle operation using NVSHMEM, with split information provided on device.
+   * Arguments:
+   *  - `input` is the input tensor
+   *  - `out` is the output tensor
+   *  - `in_out_splits` is a 2D tensor of size (3, `world_size` * `ne`). In the
+   *  scenario of Mixture-of-Experts models, `ne` is the number of experts per
+   *  rank. The rows are (in order):
+        input splits (IN)
+        output splits (OUT) and
+        output offsets (OUT).
+   *  - `group_name` is the name of the group to use for the collective operation.
+   *  A 2D AllToAllv shuffle is illustrated below:
+        (world_size = 2, ne = 2)
+        Source: |         Rank 0        |         Rank 1        |
+                | e00 | e01 | e02 | e03 | e10 | e11 | e12 | e13 |
+        Dest:   |         Rank 0        |         Rank 1        |
+                | e00 | e10 | e01 | e11 | e02 | e12 | e03 | e13 |
+        where each eij is a slice of the `input` tensor, with length indicated by input splits (`in_out_splits[0]`) at index `i * ne + j`.
+        That is, the 2D AllToAllv shuffle achives a transpose from rank-major order at input to expert-major order at output.
+  */
+  auto input_hdl = c10d::symmetric_memory::rendezvous(input, group_name);
+  auto out_hdl = c10d::symmetric_memory::rendezvous(out, group_name);
+  auto splits_hdl = c10d::symmetric_memory::rendezvous(in_out_splits, group_name);
+  int rank = input_hdl->get_rank();
+  int world_size = input_hdl->get_world_size();
+
+  void* input_ptr = input_hdl->get_buffer_ptrs()[rank];
+  void* output_ptr = out_hdl->get_buffer_ptrs()[rank];
+  int64_t* splits_ptr = (int64_t*)(splits_hdl->get_buffer_ptrs()[rank]);
+
+  // Number of experts per rank
+  int ne = in_out_splits.stride(0) / world_size;
+  TORCH_CHECK(ne * world_size == in_out_splits.stride(0), "Each row of in_out_splits must be a multiple of world_size")
+
+  auto stream = at::cuda::getCurrentCUDAStream(input.device().index());
```

Comment:
you need to check that current device matches input's device (unless device guard code is generated automatically, I don't know how you function is registered), and then just use getCurrentCUDAStream() here because you'd get stream for current device

Reply:
I'd prefer if the API do not require user setting a device guard upfront. 
So, I will do something like the following here:
```
at::cuda::OptionalCUDAGuard gpuGuard(input.device());
auto stream = at::cuda::getCurrentCUDAStream();
nvshmemx_collective_launch(..., stream);
```